README.md
### Megatron - A System for Abuse- and Incident Handling
Megatron is a tool implemented by CERT-SE which collects and analyses log files with bad 
machines, e.g. from Shadowserver. Apart from abuse mail handling, Megatron can be used to 
collect statistics, convert log files, and do log file analysis during incident handling.

Major features:
* Flexible parsing: many input sources are supported without any coding. Regular expressions are 
  used to extract tokens which are bound to variables. A variable corresponds to a field in the 
  database and can be used in for example filters and templates.
* Organization matching: Megatron tries to match every input record to an organization using its 
  IP-blocks, ASNs, and domain names. Each organization have a priority which makes it possible to 
  filter out important records.
* Database: input records are stored in a relational database (MySQL is default), making data mining possible.
* Filtering: a wide variety of filters can be used to filter input records.
* Data decoration: data may be added to the record before it is saved to the database. Types of lookups: 
  IP to ASN, IP to country code, IP to hostname, hostname to IP, URL to hostname, hostname to 
  country code, and IP to geolocation (longitude, latitude, and city). Three of MaxMind's databases
  are supported (ASN, City, and Country).
* Performance: large data volumes can be handled. All lookups except DNS queries are local, for instance 
  ASN lookups uses database queries by importing a BGP routing table to the database.

For all features, see [readme-general.txt](https://github.com/cert-se/megatron-java/blob/master/doc/readme-general.txt)


### Example of Usage
Convert a file with IP addresses to a pipe-separated file with IP, AS, country code, and hostname:
```      
$ megatron.sh --job-type ip-flowing --export --no-db test-data/multiple-ips-per-line3.log
```

As above but output file is tab-separated and contains geolocation data:
```
$ megatron.sh --job-type ip-flowing-verbose --export --no-db test-data/multiple-ips-per-line3.log
```

Prints information about two IP addresses (hostnames or URLs works as well):
```
$ megatron.sh --whois 130.238.7.133 217.21.237.35
IP              | AS     | CC | Hostname                                      | AS Name                                       | Organization
130.238.7.133   | 1653   | SE | live.webb.uu.se                               | SUNET Swedish University Network              | Uppsala universitet
217.21.237.35   | 29672  | SE | stockholm.se                                  | S:t Erik Kommunikation AB                     | Stockholms stad
```

Prints information about URLs in specified file:
```
$ megatron.sh --whois infection-urls.txt
```

Process file and save result in the database:
```
$ megatron.sh --job-type shadowserver-drone test-data/2009-06-08-drone-report-se.log
```

Preview of mail to be sent:
```
$ megatron.sh --job shadowserver-drone_2013-06-22_160142 --id 4242 --mail-dry-run
```

Send emails for the job:
```
$ megatron.sh --job shadowserver-drone_2009-06-22_160142 --id 4242 --mail
```

Megatron is designed to automate day-to-day abuse handling. Running Megatron from the
command line is only necessary in special cases.  


### Download

Binaries, config files, and source code are included in this tarball:
* [megatron-pub-1.1.0.tar.gz](https://www.cert.se/megatron/megatron-pub-1.1.0.tar.gz)
<pre>MD5: 816a6b9644cc929822e861221b07ea37
SHA1: 77c8c1bedd6318574b34d85daaaae8a6d94e6b1e</pre>
* [Release Notes](https://github.com/cert-se/megatron-java/blob/master/doc/release-notes.txt)


### Installation
Megatron is cross-platform as it is implemented in Java and Python. It has been tested on Windows 7 and Ubuntu 12.04. 
Megatron requires a MySQL database. Do not worry, the installation is easy and described in detail in this document:
[readme-install.txt](https://github.com/cert-se/megatron-java/blob/master/doc/readme-install.txt)


### Documentation
To get a grasp of Megatron, we recommend skimming the following:
* Overview documentation: [readme-general.txt](https://github.com/cert-se/megatron-java/blob/master/doc/readme-general.txt)
* Database schema: [megatron-schema.sql](https://github.com/cert-se/megatron-java/blob/master/sql/megatron-schema.sql)
* Configuration files: [conf/dev/megatron-globals.properties](https://github.com/cert-se/megatron-java/blob/master/conf/dev/megatron-globals.properties) 
and [conf/job-type/*](https://github.com/cert-se/megatron-java/tree/master/conf/job-type)
* The [FAQ](https://github.com/cert-se/megatron-java/wiki/Megatron-FAQ)
* The source code

Other resources are [this presentation](https://www.cert.se/megatron/megatron-telia2011.pdf) and the [projects Wiki](https://github.com/cert-se/megatron-java/wiki).


### License
Megatron is distributed under the terms of the Apache Software Foundation license version 2.0, which is included in the 
file LICENSE in the root of the project.


### Support and Warranty
CERT-SE does not offer any support, and Megatron is provided "as is" without warranty of any kind.

However, we are a bunch of nerds at CERT-SE and we would love to hear from you and discuss technical stuff. 
Send feedback to <cert@cert.se>. Our PGP keys can be found on CERT-SE's [contact page](https://www.cert.se/om-cert-se).


### Mailing List
[Megatron Hacking](https://groups.google.com/group/megatron-hacking) is the mailing list for Megatron. Join by sending an empty email to:

  <megatron-hacking+subscribe@googlegroups.com>
  
Or you can go to the following URL:

  <https://groups.google.com/group/megatron-hacking/subscribe>

The mailing list is for discussions about Megatron (installation problems, bug fixes, feature 
requests, etc.). Keep in mind that the list is public.


doc/release-notes.txt
v1.1.1
======


* Changes to the command line interface OrganizationHandler
  - Search organisations by prio number
  - List contacts for an organization
  - Possible to enable/disable contact without having go though the
    editing of all the contact attributes.
  - Fixed the problem with selection of AS numbers, IP ranges or domain names
    not being commited to the database.



v1.1.0
======

* New: Contact, new class that contains additional information about the 
  organization contact data e.g. email address, name, phone number.
  The information is stored in the contact table in the database. See the
  OrganizationHandler improvements below.

* New: OrganizationContactMigrator, helper class that is used to migrate
  the database from the previous schema to the schema in this relase. The class
  is invoked from the megatron.sh script using the --class option.
  
* New: ITicketHandler, interface for an otional ticket handler implementation.
  The implementing class should be used if interaction with a ticketing
  system is needed. The current interface has a method for getting a new ticket
  ID from the tickeing system. That ID can be used in the e-mail subject in the
  abuse notification e-mails. The implementing class in configurable and also
  if the e-mail should be sent by Megatron or through the ticketing system.
  An implementing class in not included in this release.

* Improved: OrganizationHandler, the command line interface for managing 
  organizations had been improved with several new functions and improvments to
  existing features.

  Additional features:
  - Search organizations by name, ASN, domain name, contact e-mail address,
    IP-address or IP-range.
  - Adding of organization contacts. One organization can have multiple 
    contacts. Each contact can consist of first & last name, e-mail address, 
    e-mail type (To, Cc or Bcc), contact role (configurable e.g. Abuse, 
    Administrative etc.) and an reference to an external contact database if
    that is used to maintain the organization contacts.
  - Possible to list all existing organization priority levels.
  - Search log entries by organizaton ID, ASN, IP-address or IP-range.
  - View specific log entries.
  - Export contact information (to file or screen) by organization and role or
    organization priority level.
  - Individual contacts (e-mail addesses) can be enabled/disabled.

  Improvements:
  - IP-ranges can be entered in the formats: x.x.x.x-y.y.y.y, x.x.x.x-y or 
    x.x.x.x/y when administrating organizations or searching by ranges.
  - Better error message when entrering overlaping IP-ranges.
  - Easier editing of organization search properties.

* Improved: Class can be specified in megatron.sh using --class.

* Config changes:
    Megatron properties:
    - New ui.organizationHandler.defaultCountryCode
    - New ui.organizationHandler.defaultLanguageCode
    - New ui.organizationHandler.outputDir
    - New ui.organizationHandler.validRoles
    - New ui.organizationHandler.timestampFormat
    - New ticketHandler.className
    - New ticketHandler.createChild
    - New ticketHandler.valueKeys
    - New ticketHandler.sendsMail
    - New ticketHandler.queueName
    - New ticketHandler.user
    - New ticketHandler.password
    - New ticketHandler.ticketOwner
    - New ticketHandler.url

    Hibernate mapping:
    - New: Contact.hbm.xml 
      - new table contact and class Contact.
    - Changed: Organization.hbm.xml
      - removed EmailAddresses and AutoUpdateEmail
      - added set of Contacts

    Hibernate config:
    - Changed: hibernate.cfg.xml
           - added mapping resource Contact.hbm.xml 

* Database schema changes:
  - Changed: megatron-schema.sql:
    - New table contact, see file sql/megatron-schema.sql for details.
    - Moved column auto_update_email from organization table to contact table.
    - Removed column email_addresses in organization table

v1.0.12
=======
* Bug: ASN in log record is not matched with organization when 
  "decorator.asnGeoIpDecorator.useAsnInLogEntry=false".


v1.0.11
=======
* New: The following job types (configurations) are new:
    - ikyon
    - brobot
    - brobot2
    - ddos-amplification
    - ip-flowing-fast
    - ip-flowing-verbose
    - megatron-whois-ip
    - megatron-whois-hostname
    - shadowserver-openresolver
    - shadowserver-qotd
    - shadowserver-chargen
    - shadowserver-sinkhole-http-drone

src/Megatron.java
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import org.apache.log4j.Logger;
import org.apache.log4j.PropertyConfigurator;

import se.sitic.megatron.core.AppProperties;
import se.sitic.megatron.core.CommandLineParseException;
import se.sitic.megatron.core.EmailAddressBatchUpdater;
import se.sitic.megatron.core.JobInfoWriter;
import se.sitic.megatron.core.JobListWriter;
import se.sitic.megatron.core.JobScheduler;
import se.sitic.megatron.core.MegatronException;
import se.sitic.megatron.core.NetnameUpdater;
import se.sitic.megatron.core.StatsRssGenerator;
import se.sitic.megatron.core.TypedProperties;
import se.sitic.megatron.core.WhoisWriter;
import se.sitic.megatron.db.DbManager;
import se.sitic.megatron.db.ImportBgpTable;
import se.sitic.megatron.db.ImportSystemData;
import se.sitic.megatron.entity.Job;
import se.sitic.megatron.entity.Priority;
import se.sitic.megatron.report.IReportGenerator;
import se.sitic.megatron.ui.OrganizationHandler;
import se.sitic.megatron.util.Constants;
import se.sitic.megatron.util.FileUtil;
import se.sitic.megatron.util.StringUtil;
import se.sitic.megatron.util.Version;


/**
 * Main class for Megatron. This class is called from the CLI.
 */
public class Megatron {
    /** Key in system properties for log4j config file. */
    private static final String LOG4J_FILE_KEY = "megatron.log4jfile";

    // Note:
    // * --overwrite is not used. use --delete plus re-run instead.
    // * --export may be used with --output-dir or use filename in config
    // * Log job + mail job is not allowed, because it breaks preview.
    // * --mail requires --id plus and job. --no-db is not allowed.
    private static final String USAGE =
        "_______ _______  ______ _______ _______  ______  _____  __   _" + Constants.LINE_BREAK +
        "|  |  | |______ |  ____ |_____|    |    |_____/ |     | | \\  |" + Constants.LINE_BREAK +
        "|  |  | |______ |_____| |     |    |    |    \\_ |_____| |  \\_|" + Constants.LINE_BREAK +
        "" + Constants.LINE_BREAK +
        "Usage: megatron.sh [options] logfiles(s)" + Constants.LINE_BREAK +
        "" + Constants.LINE_BREAK +
        "Options:" + Constants.LINE_BREAK +
        "  -v, --version        Print application version and exit." + Constants.LINE_BREAK +
        "  -h, --help           Print this help message." + Constants.LINE_BREAK +
        "  -s, --slurp          Process files in the slurp directory." + Constants.LINE_BREAK +
        "  -l, --list-jobs      List processed log jobs. No. of days may be specified." + Constants.LINE_BREAK +
        "  -w, --whois          Print whois report for specified IPs or hostnames." + Constants.LINE_BREAK +
        "  -e, --export         Export log records to file." + Constants.LINE_BREAK +
        "  -d, --delete         Delete job including log records." + Constants.LINE_BREAK +
        "  -D, --delete-all     Delete job plus mail jobs." + Constants.LINE_BREAK +
        "  -j, --job            Specifies job, e.g. 'shadowserver-drone_2009-06-22_084510'." + Constants.LINE_BREAK +
        "  -t, --job-type       Specifies job type for input files, e.g. 'shadowserver-drone'." + Constants.LINE_BREAK +
        "  -o, --output-dir     Specifies directory for export files." + Constants.LINE_BREAK +
        "  -i, --id             Specifies RTIR id." + Constants.LINE_BREAK +
        "  -p, --prio           Specifies priority for log records to be emailed or exported." + Constants.LINE_BREAK +
        "  -P, --list-prios     List priorities." + Constants.LINE_BREAK +
        "  -I, --job-info       Print info about specified log job." + Constants.LINE_BREAK +
        "  -n, --no-db          Skip writes to the database."  + Constants.LINE_BREAK +
        "  -S, --stdout         Writes to stdout instead of export file."  + Constants.LINE_BREAK +
        "  -1, --mail-dry-run   Create a mail report but does not send any mail." + Constants.LINE_BREAK +
        "  -2, --mail-dry-run2  As '--mail-dry-run' but more verbose." + Constants.LINE_BREAK +
        "  -m, --mail           Send mails for a job."  + Constants.LINE_BREAK +
        "  -b, --use-org2       Use secondary organization when mailing." + Constants.LINE_BREAK +
        "" + Constants.LINE_BREAK +
        "Admin Options:" + Constants.LINE_BREAK +
        "  --import-contacts    Import organizations to the database." + Constants.LINE_BREAK +
        "  --import-bgp         Import BGP dump file (specified in config)." + Constants.LINE_BREAK +
        "  --update-netname     Update netname field from whois queries." + Constants.LINE_BREAK +
        "  --add-addresses      Add email addresses listed in specified file." + Constants.LINE_BREAK +
        "  --delete-addresses   Delete email addresses listed in specified file." + Constants.LINE_BREAK +
        "  --create-rss         Create RSS with Megatron statistics."+ Constants.LINE_BREAK +
        "  --create-reports     Create report files (json, xml, html, etc.)."+ Constants.LINE_BREAK +
        "  --create-report      Run a specific report."+ Constants.LINE_BREAK +
        "  --ui-org             Administration of organizations (command line interface)." + Constants.LINE_BREAK +
        "" + Constants.LINE_BREAK +
        "Examples:" + Constants.LINE_BREAK +
        "  Process file and save result in the database:" + Constants.LINE_BREAK +
        "    megatron.sh --job-type shadowserver-drone 2009-06-22-drone-report-se.csv" + Constants.LINE_BREAK +
        "  Preview of mail to be sent:" + Constants.LINE_BREAK +
        "    megatron.sh --job shadowserver-drone_2009-06-22_160142 --id 4242 --mail-dry-run" + Constants.LINE_BREAK +
        "  Send mails for the job:" + Constants.LINE_BREAK +
        "    megatron.sh --job shadowserver-drone_2009-06-22_160142 --id 4242 --mail" + Constants.LINE_BREAK +
        "  As above, but sends only to organizations with a prio of 50 or above:" + Constants.LINE_BREAK +
        "    megatron.sh --job shadowserver-drone_2009-06-22_160142 --id 4242 --prio 50 --mail" + Constants.LINE_BREAK +
        "  Display information, e.g. high priority organizations, about specified job:" + Constants.LINE_BREAK +
        "    megatron.sh --job-info shadowserver-drone_2009-06-15_124508" + Constants.LINE_BREAK +
        "  Delete specified job:" + Constants.LINE_BREAK +
        "    megatron.sh --delete shadowserver-drone_2009-06-15_124508" + Constants.LINE_BREAK +
        "  Display log jobs created the last 4 days:" + Constants.LINE_BREAK +
        "    megatron.sh --list-jobs 4" + Constants.LINE_BREAK +
        "  Process files and exports them to a different format:" + Constants.LINE_BREAK +
        "    megatron.sh --export --job-type whois-cymru-verbose --no-db file1.txt file2.txt" + Constants.LINE_BREAK +
        "  Add email addresses listed in file to the database:" + Constants.LINE_BREAK +
        "    megatron.sh --add-addresses new-addresses.txt" + Constants.LINE_BREAK +
        "  Start admin-UI for organizations, IP-blocks, ASNs, and domain names:" + Constants.LINE_BREAK +
        "    megatron.sh --ui-org" + Constants.LINE_BREAK +
        "  Print whois report for specified IPs (hostnames or URLs will also work):" + Constants.LINE_BREAK +
        "    megatron.sh --whois 192.121.218.90 1.1.1.1 2.2.2.2 8.8.8.8" + Constants.LINE_BREAK +
        "  Print whois report for specified file with IPs, hostnames, or URLs:" + Constants.LINE_BREAK +
        "    megatron.sh --whois infected.txt" + Constants.LINE_BREAK +
        "  Run the organization report (emails an abuse-report to selected organizations):" + Constants.LINE_BREAK +
        "    megatron.sh --create-report se.sitic.megatron.report.OrganizationReportGenerator" + Constants.LINE_BREAK +
        "  Process files in the slurp directory and exports them to a different format:" + Constants.LINE_BREAK +
        "    megatron.sh --slurp --no-db --export" + Constants.LINE_BREAK +
        "  Process files in the slurp directory and save result in the database:" + Constants.LINE_BREAK +
        "    megatron.sh --slurp" + Constants.LINE_BREAK;

    private static final String CLI_ERROR =
        "@errorMsg@" + Constants.LINE_BREAK +
        "--help will show help message." + Constants.LINE_BREAK +
        "" + Constants.LINE_BREAK;


    /**
     * Main.
     */
    public static void main(String[] args) {
        Logger log = null;
        try {
            Megatron megatron = new Megatron();

            // -- Read config files and parse command line
            try {
                AppProperties.getInstance().init(args);
            } catch (CommandLineParseException e) {
                if (e.getAction() == CommandLineParseException.SHOW_USAGE_ACTION) {
                    System.out.println(USAGE);
                    System.exit(0);
                } else if (e.getAction() == CommandLineParseException.SHOW_VERSION_ACTION) {
                    System.out.println(Version.getVersionInfo());
                    System.exit(0);
                } else {
                    String errorMsg = (e.getMessage() != null) ? e.getMessage() : "";
                    String msg = StringUtil.replace(CLI_ERROR, "@errorMsg@", errorMsg);
                    System.err.println(msg);
                    System.exit(1);
                }
            } catch (MegatronException e) {
                String msg = "Error: Cannot initialize configuration: " + e.getMessage();
                System.err.println(msg);
                e.printStackTrace();
                System.exit(1);
            }
            
            // -- Init logging
            megatron.ensureDirs();
            try {
                megatron.initLogging();
            } catch (FileNotFoundException e) {
                String msg = "Error: Cannot initialize logging: " + e.getMessage();
                System.err.println(msg);
                e.printStackTrace();
                System.exit(2);
            }
            log = Logger.getLogger(Megatron.class);
            
            log.info(megatron.getVersion() + " started.");
            megatron.processCommands();
            log.info(Version.getAppName() + " finished.");
            
            System.exit(0);
        } catch (Throwable e) {
            String msg = (e instanceof MegatronException) ? "Error: " + e.getMessage() : "Error (unhandled exception): " + e.getMessage();

            if (log != null) {
                // Big Brother looks for "fatal" errors.
                log.fatal(msg, e);
            }


            // Write message to console (without a stack trace, which may scare the user)
            System.err.println(msg);
            Throwable cause = e.getCause();
            while (cause != null) {
                msg = cause.getMessage();
                if (StringUtil.isNullOrEmpty(msg) || msg.equalsIgnoreCase("null")) {
                    msg = "[Message not available]";
                }
                System.err.println("  " + msg);
                cause = cause.getCause();
            }
            try {
                String dir = AppProperties.getInstance().getGlobalProperties().getString(AppProperties.LOG_DIR_KEY, "[property missing]");
                // hardcoded; filename is specified in log4j.properties
                String filename = "megatron.log";    
                File logFile = new File(dir, filename);
                System.err.println("See log file for more info: " + logFile.getAbsoluteFile());
            } catch (Exception e2) {
                // ignored
            }

            System.exit(100);
        }
    }


    private void initLogging() throws FileNotFoundException {
        String log4jFilename = System.getProperty(LOG4J_FILE_KEY);
        String defaultFilename = AppProperties.getInstance().getGlobalProperties().getString(AppProperties.LOG4J_FILE_KEY, null);
        log4jFilename = (log4jFilename != null) ? log4jFilename : defaultFilename;
        File file = new File(log4jFilename);
        if (!file.exists()) {
            String msg = "Cannot find log4j config-file: " + file.getAbsoluteFile() + ". Working directory for Megatron is probably wrong.";
            throw new FileNotFoundException(msg);
        }

        PropertyConfigurator.configure(log4jFilename);
    }


    /**
     * Creates, if necessary, directories required by the application.
     */
    private void ensureDirs() throws IOException {
        TypedProperties globalProps = AppProperties.getInstance().getGlobalProperties();
        String dir = null; 
        dir = globalProps.getString(AppProperties.LOG_DIR_KEY, null);
        FileUtil.ensureDir(dir);
        dir = globalProps.getOutputDir();
        FileUtil.ensureDir(dir);
        dir = globalProps.getString(AppProperties.SLURP_DIR_KEY, null);
        FileUtil.ensureDir(dir);
    }


    private String getVersion() {
        return Version.getVersion(true);
    }

    
    /**
     * Process commands in the command line.
     */
    private void processCommands() throws MegatronException {
        TypedProperties globalProps = AppProperties.getInstance().getGlobalProperties();

        if (globalProps.isWhois()) {
            // inputFiles contains IPs and hostnames
            WhoisWriter writer = new WhoisWriter(globalProps, true);
            writer.execute();
            return;
        }
        
        List<String> inputFiles = AppProperties.getInstance().getInputFiles();
        if (inputFiles.size() > 0) {
            for (Iterator<String> iterator = inputFiles.iterator(); iterator.hasNext();) {
                File file = new File(iterator.next());
                String jobType = AppProperties.getInstance().mapFilenameToJobType(file.getName(), false);
                if (jobType == null) {
                    throw new MegatronException("Cannot find a job-type. Use '--job-type' to specify one.");
                }
                TypedProperties props = AppProperties.getInstance().createTypedPropertiesForCli(jobType);
                JobScheduler.getInstance().processFile(props, file);
            }
        } else if (globalProps.isListJobs()) {
            JobListWriter writer = new JobListWriter(globalProps, true);
            writer.execute();
        } else if (globalProps.isListPrios()) {
            listPrio(globalProps);
        } else if (globalProps.isJobInfo()) {
            JobInfoWriter writer = new JobInfoWriter(globalProps);
            writer.execute();
        } else if (globalProps.isSlurp()) {
            JobScheduler.getInstance().processSlurpDirectory();
        } else if (globalProps.isExport()) {
            String jobName = globalProps.getJob();
            if (jobName == null) {
                throw new MegatronException("Invalid command line: No job ('--job') specified.");
            }
            
            String jobType = getJobType(globalProps, jobName);
            TypedProperties props = AppProperties.getInstance().createTypedPropertiesForCli(jobType);
            JobScheduler.getInstance().processFileExport(props, jobName);
        } else if (globalProps.isMail() || globalProps.isMailDryRun() || globalProps.isMailDryRun2()) {
            // check CLI args
            String id = globalProps.getParentTicketId();
            if (id == null) {
                throw new MegatronException("Invalid command line: No RTIR-id ('--id') specified.");
            }
            String jobName = globalProps.getJob();
            if (jobName == null) {
                throw new MegatronException("Invalid command line: No job ('--job') specified.");
            }
            if (globalProps.isNoDb()) {
                throw new MegatronException("Invalid command line: --no-db cannot be used when mailing.");
            }
            
            // run mail job
            String jobType = getJobType(globalProps, jobName);
            TypedProperties props = AppProperties.getInstance().createTypedPropertiesForCli(jobType);
            JobScheduler.getInstance().processMailJob(props, jobName);
        } else if (globalProps.isImportContacts()) {
            ImportSystemData dataImporter = new ImportSystemData(globalProps);
            dataImporter.importFile();            
        } else if (globalProps.isImportBgp()) {
            ImportBgpTable importer = new ImportBgpTable(globalProps);
            importer.importFile();
        } else if (globalProps.isUpdateNetname()) {
            NetnameUpdater updater = new NetnameUpdater(globalProps);
            updater.update();
        } else if (globalProps.isAddAddresses()) {
            EmailAddressBatchUpdater updater = new EmailAddressBatchUpdater(globalProps);
            updater.addAddresses(globalProps.getAddressesFile());
        } else if (globalProps.isDeleteAddresses()) {
            EmailAddressBatchUpdater updater = new EmailAddressBatchUpdater(globalProps);
            updater.deleteAddresses(globalProps.getAddressesFile());
        } else if (globalProps.isCreateStatsRss()) {
            StatsRssGenerator generator = new StatsRssGenerator(globalProps);
            generator.createFile();
        } else if (globalProps.isCreateFlashXml() || globalProps.isCreateReports() || (globalProps.getCreateReport() != null)) {
            String classNames[] = null;
            if (globalProps.getCreateReport() != null) {
                classNames = new String[1];
                classNames[0] = globalProps.getCreateReport();
            } else {
                classNames = globalProps.getStringList(AppProperties.REPORT_CLASS_NAMES_KEY, new String[0]);
            }
            List<IReportGenerator> reportGenerators = createReportGenerators(classNames);
            if (reportGenerators.isEmpty()) {
                throw new MegatronException("No report generators specified. See '" + AppProperties.REPORT_CLASS_NAMES_KEY + "'." );
            }
            for (Iterator<IReportGenerator> iterator = reportGenerators.iterator(); iterator.hasNext(); ) {
                iterator.next().createFiles();
            }
        } else if (globalProps.isUiOrg()) {
            OrganizationHandler uiOrg = new OrganizationHandler(globalProps);
            uiOrg.startUI();                    
        } else if (globalProps.isDelete() || globalProps.isDeleteAll()) {
            String jobName = globalProps.getJob();
            if (jobName == null) {
                throw new MegatronException("Invalid command line: No job ('--job') specified.");
            }
            if (globalProps.isNoDb()) {
                throw new MegatronException("Invalid command line: --no-db cannot be used when deleting.");
            }

            if (globalProps.isDeleteAll()) {
                JobScheduler.getInstance().deleteAll(globalProps, jobName);
            } else {
                JobScheduler.getInstance().delete(globalProps, jobName);
            }
            System.out.println("Job deleted: " + jobName);
        } else {
            // nothing to do; show usage
            System.out.println(USAGE);
        }
    }

    
    /**
     * Fetches job type from db for specified job name.
     */
    private String getJobType(TypedProperties globalProps, String jobName) throws MegatronException {
        DbManager dbManager = null;
        try {
            dbManager = DbManager.createDbManager(globalProps);
            Job job = dbManager.searchLogJob(jobName);
            if (job == null) {
                String msg = "Cannot find job: " + jobName;
                throw new MegatronException(msg);
            }
            String result = job.getJobType().getName();
            if (Constants.DEFAULT_JOB_TYPE.equals(result)) {
                // use job type in job name if job type not specified in db  
                String[] headTail = StringUtil.splitHeadTail(jobName, "_", false);
                result = headTail[0];
            }
            return result;
        } finally {
            if (dbManager != null) {
                dbManager.close();
            }
        }
    }
    
    
    /**
     * List all priorites to stdout. 
     */
    private void listPrio(TypedProperties globalProps) throws MegatronException {
        DbManager dbManager = null;
        try {
            dbManager = DbManager.createDbManager(globalProps);
            List<Priority> prios = dbManager.getAllPriorities();
            System.out.println("+------+----------------------------------------------------------------------------+");
            System.out.println("| Prio | Name                                                                       |");                
            System.out.println("+------+----------------------------------------------------------------------------+");
            for (Priority prio: prios) {
                System.out.printf("| %4d | %-74s |\n", prio.getPrio(), prio.getName());
            }
            System.out.println("+------+----------------------------------------------------------------------------+");
        } finally {
            if (dbManager != null) {
                dbManager.close();
            }
        }
    }
    
    
    private List<IReportGenerator> createReportGenerators(String[] classNames) throws MegatronException {
        List<IReportGenerator> result = new ArrayList<IReportGenerator>();
        for (int i = 0; i < classNames.length; i++) {
            String className =  classNames[i];
            if (className.trim().length() == 0) {
                continue;
            }
            try {
                Class<?> clazz = Class.forName(className);
                IReportGenerator reportGenerator = (IReportGenerator)clazz.newInstance();
                reportGenerator.init();
                result.add(reportGenerator);
            } catch (MegatronException e) {
                String msg = "Cannot initialize report generator class: " + className;
                throw new MegatronException(msg, e);
            } catch (Exception e) {
                // ClassNotFoundException, InstantiationException, IllegalAccessException
                String msg = "Cannot instantiate report generator class: " + className;
                throw new MegatronException(msg, e);
            }
        }
        return result;
    }

}
